### PR TITLE
fix: run proto disto on epoch

### PR DESCRIPTION
**Pull request by PaddyMc**
Closes: #8089
Change from this PR: https://github.com/osmosis-labs/osmosis/pull/8054

### Needed to redo the PR from osmosis main as e2e won't work rn for external contributors

Fixes the registration of epoch hooks in protorev.   
[EpochHooks](https://github.com/osmosis-labs/osmosis/blob/fe4f0c6eaa5596edcaaca7779abab3423604ebc7/app/keepers/keepers.go#L824) are set for protorev by the `EpochHooks()` method. Those are created as an `EpochHooks` item, and not directly as part of the keeper. 

In #7564 we added hooks directly to the keeper which were never registered with the epoch hooks. Instead of having these live in separate places I've just combined the newly added logic into the existing AfterEpoch hook.

All logic remains the same.

## Testing and Verifying

Existing unit test coverage still applies to all changes--I've simply copy/pasted some of those to organize all epoch-specific logic into one place within protorev.

Not sure if you have e2e tests somewhere that I can update to test protorev distributions?

## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes?
  - [ ] Changelog entry added to `Unreleased` section of `CHANGELOG.md`?

Where is the change documented?
  - [ ] Specification (`x/{module}/README.md`)
  - [ ] Osmosis documentation site
  - [ ] Code comments?
  - [ ] N/A

## Summary by CodeRabbit

- **New Features**
	- Enabled ProtoRev distribution on epoch.
	- Added a deactivated smart account module.
- **Bug Fixes**
	- Reset validator signing info missed blocks counter.
	- Deleted legacy behavior related to lockups unbonding at small block heights on a testnet.
- **Refactor**
	- Improved profit calculation and distribution process for the ProtoRev module.
	- Restructured epoch-related hooks to enhance module efficiency.
- **Tests**
	- Updated test cases to reflect new logic in profit distribution and epoch handling.

### CODE REVIEW

**Review by PaddyMc — COMMENTED** [review]
Debugged through the epoch to ensure this code runs as expected:

```
{"level":"info","module":"server","time":"2024-04-19T10:43:55Z","message":"distributeInternal CL for pool id 1671 finished"}
{"level":"info","module":"server","time":"2024-04-19T10:43:55Z","message":"x/incentives AfterEpochEnd finished distribution"}
{"level":"info","module":"server","mint":"mintedCoins","time":"2024-04-19T10:43:55Z","message":"AfterEpochEnd, minted coins"}
{"level":"info","module":"server","module":"mint","recepientModule":"fee_collector","distributionCoin":"91324200913uosmo","height":15040473,"time":"2024-04-19T10:43:55Z","message":"distributeToModule"}
{"level":"info","module":"server","module":"mint","recepientModule":"poolincentives","distributionCoin":"36529680365uosmo","height":15040473,"time":"2024-04-19T10:43:55Z","message":"distributeToModule"}
{"level":"info","module":"server","module":"poolincentives","totalMintedAmount":"36529680369","height":15040473,"time":"2024-04-19T10:43:55Z","message":"AllocateAsset minted amount"}
> github.com/osmosis-labs/osmosis/v24/x/protorev/keeper.EpochHooks.AfterEpochEnd() /root/github.com/osmosis-labs/osmosis/x/protorev/keeper/epoch_hook.go:43 (hits goroutine(52):1 total:1) (PC: 0x37f99b2)
    38:	func (h EpochHooks) AfterEpochEnd(ctx sdk.Context, epochIdentifier string, epochNumber int64) error {
    39:		if h.k.GetProtoRevEnabled(ctx) {
    40:			switch epochIdentifier {
    41:			case "day":
    42:				// Calculate and distribute protorev profits
=>  43:				err := h.CalculateDistributeProfits(ctx)
    44:				if err != nil {
    45:					return err
    46:				}
    47:	
    48:				// Increment number of days since module genesis to properly calculate developer fees after cyclic arbitrage trades

```

File: x/protorev/keeper/epoch_hook.go [hunk 1]
```diff
@@ -54,6 +60,47 @@ func (h EpochHooks) AfterEpochEnd(ctx sdk.Context, epochIdentifier string, epoch
 	return nil
 }
 
+// CalculateDistributeProfits is executed after epoch. It gets the current base denom profits and distributes them.
+func (h EpochHooks) CalculateDistributeProfits(ctx sdk.Context) error {
```
Comment (nicolaslara) [inline x/protorev/keeper/epoch_hook.go:64]: Nit: the current name makes me think this function has no side-effects. Renaming makes it clearer that distributing happens here

```suggestion
func (h EpochHooks) CalculateAndDistributeProfits(ctx sdk.Context) error {
```

File: x/protorev/keeper/epoch_hook.go [hunk 2]
```diff
@@ -39,6 +39,12 @@ func (h EpochHooks) AfterEpochEnd(ctx sdk.Context, epochIdentifier string, epoch
 	if h.k.GetProtoRevEnabled(ctx) {
 		switch epochIdentifier {
 		case "day":
+			// Calculate and distribute protorev profits
+			err := h.CalculateDistributeProfits(ctx)
```
Comment (nicolaslara) [inline x/protorev/keeper/epoch_hook.go:43]: Renaming here too to match the other rename
```suggestion
			err := h.CalculateAndDistributeProfits(ctx)
```

**Review by nicolaslara — APPROVED** [review]

File: x/protorev/keeper/epoch_hook.go (hunk 2 above)
Comment (PaddyMc) [reply x/protorev/keeper/epoch_hook.go]: Good idea updated here => https://github.com/osmosis-labs/osmosis/pull/8106/commits/e043b86ebc451eb183f8de6fc70e26955c1fb29d

File: x/protorev/keeper/epoch_hook.go (hunk 1 above)
Comment (PaddyMc) [reply x/protorev/keeper/epoch_hook.go]: Good idea updated here => https://github.com/osmosis-labs/osmosis/pull/8106/commits/e043b86ebc451eb183f8de6fc70e26955c1fb29d

**Pull request merged**